### PR TITLE
ENH Cache result of GridFieldDetailForm_ItemRequest::getGridFieldItemAdjacencies()

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/11703

This cuts down 10 queries to 2 when viewing a GridField managed record. This functionality being cached relates to the prev/next buttons in the bottom right of the CMS

When viewing a Member record in SecurityAdmin, the following queries each come up 5 times, have been reduced to 1 time each now

SELECT DISTINCT count(DISTINCT "Member"."ID") AS "Count" FROM "Member"
(from GridFieldPaginator::getManipulatedData():137 triggered by $this->getGridField()->getManipulatedList())

SELECT "Member"."ID", "Member"."Surname" AS "_SortColumn0", "Member"."FirstName" AS "_SortColumn1" FROM "Member" ORDER BY "_SortColumn0" ASC, "_SortColumn1" ASC LIMIT 32
(from the follow up call on the manipulated list i.e. ->column('ID'))

### CODE REVIEW

**Comment by GuySartorelli** [comment]
Assuming a hosting environment in AWS with RDS with its worst-case latency of 3ms per query, at 8 queries per gridfield, we're saving 24ms per gridfield which is juuuust about noticeably on its own.

It's not unheard of to have upwards of 10 gridfields in a single edit form, and if we assume developers aren't [setting their gridfield tabs to lazyload](https://docs.silverstripe.org/en/5/developer_guides/customising_the_admin_interface/cms_architecture/#lazy-loading-fields-on-tab-activation), that could give ~240ms per edit form load, which is well worth it especially when compounding with other performance improvements.

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 1]
```diff
@@ -627,18 +632,22 @@ public function getEditLink($id)
...
     private function getGridFieldItemAdjacencies(): array
     {
-        $list = $this->getGridField()->getManipulatedList();
-        $paginator = $this->getGridFieldPaginatorState();
-        if (!$paginator) {
-            return [];
+        if (!isset($this->cachedGridFieldItemAdjacencies)) {
+            $paginator = $this->getGridFieldPaginatorState();
+            if (!$paginator) {
+                $this->cachedGridFieldItemAdjacencies = [];
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridFieldDetailForm_ItemRequest.php]:
```suggestion
                return [];
```
Caching this empty result doesn't actually improve anything, so we may as well directly return it.

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 2]
```diff
@@ -627,18 +632,22 @@ public function getEditLink($id)
...
+            } else {
+                $currentPage = $paginator->getData('currentPage');
+                $itemsPerPage = $paginator->getData('itemsPerPage');
+                $limit = $itemsPerPage + 2;
+                $limitOffset = max(0, $itemsPerPage * ($currentPage - 1) -1);
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridFieldDetailForm_ItemRequest.php]:
```suggestion
                $limitOffset = max(0, $itemsPerPage * ($currentPage - 1) - 1);
```
Since we're adding space around the other `-` we should do it to this one.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
There are three public methods (I'll ignore the protected methods for now) that developers can call which trigger this caching:

- `ItemEditForm()`
- `getPreviousRecordID()`
- `getNextRecordID()`

If it was just `ItemEditForm()` I'd say we're probably okay to just do this as-is, but developers could for example manipulate the gridfield pagination state to set to some specific record, then use the next and previous record IDs to get the adjacent records for some purpose, then change the state again etc.

This is an edge case, but since this is a patch and not a minor we should assume that edge case is in use by somebody somewhere and not break that scenario.

The best solution I can think of is this:

1. Implement the `SilverStripe\Core\Resettable` interface. In `reset()`, unset `cachedGridFieldItemAdjacencies` or set it to null. This also makes testing easier as we now have a method to call to reset the cache state.
2. At the start of the above mentioned public methods, call the `reset()` method. This should give the same reduction in db queries mentioned in the PR description without risking breaking edge cases.
3. Add unit tests to ensure the above works as expected.

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php (hunk 1 above)
Comment (emteknetnz) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php]: Updated

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php (hunk 2 above)
Comment (emteknetnz) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php]: Updated

**Comment by emteknetnz** [comment]
This is targetting 5.4 so we cannot add any new API, so the resettable idea won't work

Instead I've simply added `$key = md5("$limit,$limitOffset");` which should take care of the case where the state is programmatically changed

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Can you please add at least one test case for this new caching?

**Comment by GuySartorelli** [comment]
Sorry, I realised in the shower this morning that even with a key based on the pagination state, this can still break edge cases. e.g. someone changes the sort of the list or the contents of the list between calling the method.

If you don't want to add the `Resettable` interface and retarget to 6.1, then please instead create a new `private` method for resetting the cache, and reset at the points mentioned in https://github.com/silverstripe/silverstripe-framework/pull/11711#pullrequestreview-2831525964

**Comment by emteknetnz** [comment]
I've updated the key to use the entire gridfield state, not just the paginator

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 3]
```diff
@@ -627,18 +632,30 @@ public function getEditLink($id)
...
+        if (!$paginatorState) {
             return [];
         }
-        $currentPage = $paginator->getData('currentPage');
-        $itemsPerPage = $paginator->getData('itemsPerPage');
-
+        $keyStr = '';
+        $data = $this->getGridField()->getState()->toArray();
+        array_walk_recursive($data, function ($v, $k) use (&$keyStr) {
+            if (is_scalar($k) && is_scalar($v)) {
```
Comment (emteknetnz) [inline src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:643]: Checking for scalars here rather than just using serialize() for the whole thing just in case something in there won't serialize properly

**Comment by GuySartorelli** [comment]
> I've updated the key to use the entire gridfield state, not just the paginator

Gridfield state doesn't know what's in the list or how it's sorted. Please use a method to reset the cache.

**Comment by emteknetnz** [comment]
I'm not convinced we need to reset the cache. This caching is on `GridFieldDetailForm_ItemRequest` i.e. the request handler for an individual item in a gridfield which is used to create an edit form. If a developer is modifying the gridfield list at this point it seems like they're doing something wrong? The caching is only for the adjacent records to the current record, and is only used for next/prev buttons. We could probably even just get rid of the cache key since they cache only exists for the duration of the request

I have tried adding the cache reset locally, adding the cache clearing to `ItemEditForm()` does make not differnece to query count, though adding it to `getPreviousRecordID()` and `ItemEditForm()` does increase it by 4 queries on a gridfield with 3 records when viewing the 2nd record.  That's obviously a tiny number of queries to increase by, though like I say I feel like cache clearing is just kind of not required here.

**Comment by GuySartorelli** [comment]
I think we need to reset the cache because, as I've mentioned before, failing to reset the cache breaks a use case that is currently valid.

> If a developer is modifying the gridfield list at this point it seems like they're doing something wrong?

I don't think we can say that, without knowing why they're doing it. And we can't know if someone is doing something, let alone why, unless someone tells us. We should assume someone is doing it and they have a valid reason to do so, since the methods are all public and there's nothing in the PHPDoc telling them they mustn't manipulate the list after calling those methods.

> The caching is only for the adjacent records to the current record, and is only used for next/prev buttons.

It's used for public methods - those could be called for reasons beyond just rendering the previous/next buttons. We shouldn't change the expected result of calling a public method.

If this was to fix a bug, then I would say it would be _maybe_ okay to do that in a minor release, where people can be warned by the changelog that something is changing. But because this is a patch and _doesn't fix a bug_, we should be risk averse and not break things.

**Comment by emteknetnz** [comment]
I just don't like this increase in complexity. Having some seemingly random reset() method to clear some tiny obscure cache for this thing that is used to calculate next/prev buttons ... because ... :shrug: It's just API bloat IMO.

At this point I think I'd rather just close this PR

If we aim this as is at `6` instead would that be sufficient?

**Comment by GuySartorelli** [comment]
Aiming at `6.1` doesn't stop a breaking change from being a breaking change. I don't think this is much complexity or really "bloat":

```php
$this->cachedGridFieldItemAdjacencies = [];
```

Doesn't even really need to be in a method, given it's a single line, just pop that in three places.

Ultimately if you don't want to persue this performance enhancement for that line, that's up to you.

**Comment by emteknetnz** [comment]
For some reason I still had the Resettable interface in my head, yes private method is fine

lol

Have added

**Review by GuySartorelli — APPROVED** [review]
LGTM
